Helper/Discount.php

```diff
@@ -0,0 +1,320 @@
+        ThirdPartyModuleFactory $amastyQuoteResource,
+        ThirdPartyModuleFactory $amastyQuoteRepository,
+        CartRepositoryInterface $quoteRepository,
+        ConfigHelper $configHelper
+    ) {
+        parent::__construct($context);
+        $this->resource = $resource;
+        $this->amastyAccountFactory = $amastyAccountFactory;
+        $this->amastyGiftCardManagement = $amastyGiftCardManagement;
+        $this->amastyQuoteFactory = $amastyQuoteFactory;
+        $this->amastyQuoteResource = $amastyQuoteResource;
+        $this->amastyQuoteRepository = $amastyQuoteRepository;
+        $this->quoteRepository = $quoteRepository;
+        $this->configHelper = $configHelper;
+    }
+
+    /**
+     * Collect and update quote totals.
+     * @param Quote $quote
+     */
+    protected function updateTotals(Quote $quote)
+    {
+        $quote->getShippingAddress()->setCollectShippingRates(true);
+        $quote->setTotalsCollectedFlag(false);
+        $quote->collectTotals();
+        $quote->setDataChanges(true);
+        $this->quoteRepository->save($quote);
+    }
+
+    /**
+     * Load Amasty Gift Card account object.
+     * @param string $code Gift Card coupon code
+     * @return \Amasty\GiftCard\Model\Account|null
+     */
+    public function loadAmastyGiftCard($code)
+    {
+        try {
+            $accountModel = $this->amastyAccountFactory->getInstance()
+                ->create()
+                ->loadByCode($code);
+            return $accountModel && $accountModel->getId() ? $accountModel : null;
+        } catch (\Exception $e) {
+            return null;
+        }
+    }
+
+    /**
+     * Apply Amasty Gift Card coupon to cart
+     *
+     * @param string $code Gift Card coupon code
+     * @param \Amasty\GiftCard\Model\Account $accountModel
+     * @param Quote $quote
+     * @return float
+     * @throws LocalizedException
+     */
+    public function applyAmastyGiftCard($code, $accountModel, $quote)
+    {
+        // Get current gift card balance before applying it to the quote
+        // in case "fixed_amount" / "pay for everything" discount type is used
+        $giftAmount = $accountModel->getCurrentValue();
+
+        $quoteId = $quote->getId();
+
+        $isValid = $this->amastyGiftCardManagement->getInstance()->validateCode($quote, $code);
+
+        if (!$isValid) {
+            throw new LocalizedException(__('Coupon with specified code "%1" is not valid.', $code));
+        }
+
+        if ($accountModel->canApplyCardForQuote($quote)) {
+            $quoteGiftCard = $this->amastyQuoteFactory->getInstance()->create();
+            $this->amastyQuoteResource->getInstance()->load($quoteGiftCard, $quoteId, 'quote_id');
+            $subtotal = $quoteGiftCard->getSubtotal($quote);
+
+            if ($quoteGiftCard->getCodeId() && $accountModel->getCodeId() == $quoteGiftCard->getCodeId()) {
+
+                throw new LocalizedException(__('This gift card account is already in the quote.'));
```

Comment:
if same code is applied, can we return the same discount information? It's possible bolt send the same request multiple times

Reply:
@daisy1754 sending the same request multiple times is taken care of in DiscountCodeValidation class by simply removing the discount (deleting a record from the db table) and applying it again. So the same data is returned.